protobuf/src/main/scala/magnolify/protobuf/ProtobufType.scala

```diff
@@ -0,0 +1,171 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package magnolify.protobuf
+
+import java.nio.ByteBuffer
+import java.{util => ju}
+
+import com.google.protobuf.Descriptors.FieldDescriptor
+import com.google.protobuf.Message
+import magnolia._
+import magnolify.shared.Converter
+import magnolify.shims.FactoryCompat
+import magnolify.shims.JavaConverters._
+
+import scala.annotation.implicitNotFound
+import scala.language.experimental.macros
+import scala.reflect.ClassTag
+
+sealed trait ProtobufType[T, ParentMsgT <: Message] extends Converter[T, ParentMsgT, ParentMsgT] {
+  def apply(r: ParentMsgT): T = from(r)
+  def apply(t: T): ParentMsgT = to(t)
+}
+
+object ProtobufType {
+  implicit def apply[T, ParentMsgT <: Message: ClassTag](
+    implicit f: ProtobufField.Record[T]
+  ): ProtobufType[T, ParentMsgT] =
+    new ProtobufType[T, ParentMsgT] {
+      override def from(v: ParentMsgT): T = f.from(v)
+      override def to(v: T): ParentMsgT =
+        f.to(
+            v,
+            implicitly[ClassTag[ParentMsgT]].runtimeClass
+              .getMethod("newBuilder")
+              .invoke(null)
+              .asInstanceOf[Message.Builder]
+          )
+          .asInstanceOf[ParentMsgT]
+    }
+}
+
+sealed trait ProtobufField[T] extends Serializable { self =>
+  type FromT
+  type ToT
+
+  def from(v: FromT): T
+  def to(v: T, b: Message.Builder): ToT
+
+  def fromAny(v: Any): T = from(v.asInstanceOf[FromT])
+}
+
+object ProtobufField {
+
+  trait Aux[T, From, To] extends ProtobufField[T] {
+    override type FromT = From
+    override type ToT = To
+  }
+
+  trait Record[T] extends Aux[T, Message, Message]
+
+  //////////////////////////////////////////////////
+
+  type Typeclass[T] = ProtobufField[T]
+
+  def combine[T](caseClass: CaseClass[Typeclass, T]): Record[T] = new Record[T] {
+
+    override def from(v: Message): T = {
+      caseClass.construct { p =>
+        val fieldDescriptor = v.getDescriptorForType.findFieldByName(p.label)
+
+        if (fieldDescriptor.getType == FieldDescriptor.Type.MESSAGE) {
+          // nested
+          p.typeclass.fromAny(v.getField(fieldDescriptor))
+        } else {
+          // non-nested
+          p.typeclass.fromAny(v.getField(fieldDescriptor))
+        }
+      }
+    }
+
+    override def to(v: T, bu: Message.Builder): Message = {
+
+      // clear builder from previous runs before using it to construct a new instance
+      caseClass.parameters
+        .foldLeft(bu.clear()) { (b, p) =>
```

Comment:
This feels fishy. Looks like all calls to this method are sharing the same builder instance created at https://github.com/spotify/magnolify/pull/51/files#diff-aa33fce540e8b204301cecff009d7a0aR47

You can probably fix this by making `bu` a `() => Message.Builder`.

Reply:
It was a deliberate choice to not have to do the (slow?) reflection within every step. I should probably do it both ways and benchmark, I will make the rest of the fixes and then look into that.

Reply:
Was gonna ask about reflection & performance next ;)
Problem with mutable state like this is potential concurrency bugs, e.g. when the same `ProtobufType` is shared by multiple threads.

To reduce reflection overhead you can probably cache the reflected method from `implicitly[ClassTag[ParentMsgT]].runtimeClass.getMethod("newBuilder")`.

Reply:
it sounds like this is probably the 'wrong way' to do it because of concurrency regardless of performance, then, thanks! I will find a way to update that. I'm also still not entirely sure what's required in terms of benchmarking though, advice welcome.

Reply:
it looks like I can't cache the reflected method at the object level because `Method` isn't serializable. I can create a new builder on this line from the existing one, instead of clearing the existing one. I'll push that, let me know what you think.

Reply:
There should be ways around serialization issue but we don't have to prematurely optimize without benchmarks, which you can plugin here: https://github.com/spotify/magnolify/blob/master/jmh/src/test/scala/magnolify/jmh/MagnolifyBench.scala